dex/src/main/scala/com/wavesplatform/dex/api/MatcherWebSocketRoute.scala

```diff
@@ -36,54 +44,91 @@ case class MatcherWebSocketRoute(addressDirectory: ActorRef,
+  private def accountUpdatesSource(publicKey: PublicKey): Source[TextMessage.Strict, ConnectionSource] = {
     Source
-      .actorRef[WsAddressState](
+      .actorRef[WsMessage](
         completionMatcher,
         failureMatcher,
         10,
         OverflowStrategy.fail
       )
-      .map(wsAddressState => TextMessage.Strict(WsAddressState.format.writes(wsAddressState).toString))
+      .map(_.toStrictTextMessage)
       .mapMaterializedValue { sourceActor =>
-        addressDirectory.tell(AddressDirectory.Envelope(publicKey, AddressActor.AddWsSubscription), sourceActor)
-        sourceActor
+        val connectionId = UUID.randomUUID()
+        addressDirectory.tell(AddressDirectory.Envelope(publicKey, AddressActor.AddWsSubscription(connectionId)), sourceActor)
+        ConnectionSource(connectionId, sourceActor)
       }
       .watchTermination()(handleTermination)
   }
 
-  private def orderBookUpdatesSource(pair: AssetPair): Source[TextMessage.Strict, Unit] =
+  private def orderBookUpdatesSource(pair: AssetPair): Source[TextMessage.Strict, ConnectionSource] =
     Source
-      .actorRef[WsOrderBook](
+      .actorRef[WsMessage](
         completionMatcher,
         failureMatcher,
         10,
         OverflowStrategy.fail
       )
-      .map(wsOrderBookState => TextMessage.Strict(WsOrderBook.wsOrderBookStateFormat.writes(wsOrderBookState).toString))
+      .map(_.toStrictTextMessage)
       .mapMaterializedValue { sourceActor =>
-        matcher.tell(MatcherActor.AddWsSubscription(pair), sourceActor)
-        sourceActor
+        val connectionId = UUID.randomUUID()
+        matcher.tell(MatcherActor.AddWsSubscription(pair, connectionId), sourceActor)
+        ConnectionSource(connectionId, sourceActor)
       }
       .watchTermination()(handleTermination)
 
-  private def createStreamFor(source: Source[TextMessage.Strict, Unit]): Route = {
-    handleWebSocketMessages(Flow.fromSinkAndSourceCoupled(Sink.ignore, source))
+  private def createStreamFor(source: Source[TextMessage.Strict, ConnectionSource]): Route = {
+
+    import webSocketSettings._
+
+    val (connectionSource, matSource) = source.preMaterialize()
+    val sinkSettings                  = SystemMessagesHandlerActor.Settings(systemMessagesSettings.pingInterval, systemMessagesSettings.pongTimeout)
+    val systemMessagesHandler         = mat.system.actorOf(SystemMessagesHandlerActor.props(sinkSettings, maxConnectionLifetime, connectionSource))
+    val sinkActor                     = Sink.actorRef(ref = systemMessagesHandler, onCompleteMessage = (), onFailureMessage = _ => Status.Failure(_))
+
+    def failed(msg: String): Future[PingOrPong]            = Future.failed[PingOrPong] { new IllegalArgumentException(msg) }
+    def binaryMessageUnsupported: Future[PingOrPong]       = failed("Binary messages are not supported")
+    def unexpectedMessage(msg: String): Future[PingOrPong] = failed(s"Got unexpected message instead of pong: $msg")
```

Comment:
Could be outside a method. `binaryMessageUnsupported` could be a `val`.

Reply:
Done